File: packages/alarm/android/src/main/kotlin/com/gdelataillade/alarm/models/NotificationSettings.kt
package com.gdelataillade.alarm.models

import android.graphics.Color
import com.gdelataillade.alarm.generated.NotificationSettingsWire
import io.flutter.Log
import kotlinx.serialization.Serializable
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * Fork addition: a custom action button rendered on the alarm notification.
 *
 * Tapping a button stops the alarm and records [id] so the Flutter app can act
 * on it (e.g. mark a medication as taken). See `NotificationActionStore`.
 */
@Serializable
data class NotificationActionButton(
    val id: String,
    val text: String,
)

@Serializable
data class NotificationSettings(
    val title: String,
    val body: String,
    val stopButton: String? = null,
    val icon: String? = null,
    val iconColor: Int? = null,
    // Fork addition. When non-empty these replace [stopButton] in the
    // notification. Defaulted so it survives the JSON round-trip through the
    // alarm Intent unchanged.
    val actionButtons: List<NotificationActionButton> = emptyList(),
) {
    companion object {
        private const val TAG = "NotificationSettings"

        // Lenient parser for the action-buttons wire payload.
        private val json = Json { ignoreUnknownKeys = true }

        fun fromWire(e: NotificationSettingsWire): NotificationSettings {
            val a = e.iconColorAlpha?.toFloat()
            val r = e.iconColorRed?.toFloat()
            val g = e.iconColorGreen?.toFloat()
            val b = e.iconColorBlue?.toFloat()

            var iconColor: Int? = null
            if (a != null && r != null && g != null && b != null) {
                iconColor = Color.argb(a, r, g, b)
            }

            // `actionButtons` arrives as a JSON-encoded string so the Pigeon
            // wire type only needs one extra trailing field.
            val actionButtons: List<NotificationActionButton> =
                e.actionButtons?.takeIf { it.isNotBlank() }?.let { raw ->
                    try {
                        json.decodeFromString(
                            ListSerializer(NotificationActionButton.serializer()),
                            raw,
                        )
                    } catch (ex: Exception) {
                        Log.e(TAG, "Failed to parse actionButtons: ${ex.message}")
                        emptyList()
                    }
                } ?: emptyList()

            return NotificationSettings(
                e.title,
                e.body,
                e.stopButton,
                e.icon,
                iconColor,
                actionButtons,
            )
        }
    }
}

File: packages/alarm/lib/model/notification_settings.dart
import 'dart:convert';
import 'dart:ui';

import 'package:alarm/model/notification_action_button.dart';
import 'package:alarm/src/generated/platform_bindings.g.dart';
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

export 'package:alarm/model/notification_action_button.dart';

part 'notification_settings.g.dart';

/// Model for notification settings.
@JsonSerializable()
class NotificationSettings extends Equatable {
  /// Constructs an instance of `NotificationSettings`.
  ///
  /// Open PR if you want more features.
  const NotificationSettings({
    required this.title,
    required this.body,
    this.stopButton,
    this.icon,
    this.iconColor,
    this.actionButtons = const [],
  });

  /// Converts the JSON object to a `NotificationSettings` instance.
  factory NotificationSettings.fromJson(Map<String, dynamic> json) =>
      _$NotificationSettingsFromJson(json);

  /// Title of the notification to be shown when alarm is triggered.
  final String title;

  /// Body of the notification to be shown when alarm is triggered.
  final String body;

  /// The text to display on the stop button of the notification.
  ///
  /// Won't work on iOS if app was killed.
  /// If null, button will not be shown. Null by default.
  ///
  /// **Fork note:** when [actionButtons] is non-empty it takes precedence on
  /// Android and [stopButton] is ignored. Keep [stopButton] set as well so iOS
  /// (which ignores [actionButtons]) still shows a usable button.
  final String? stopButton;

  /// The icon to display on the notification.
  ///
  /// **Only customizable for Android. On iOS, it will use app default icon.**
  ///
  /// This refers to the small icon that is displayed in the
  /// status bar and next to the notification content in both collapsed
  /// and expanded views.
  ///
  /// Note that the icon must be monochrome and on a transparent background and
  /// preferably 24x24 dp in size.
  ///
  /// **Only PNG and XML formats are supported at the moment.
  /// Please open an issue to request support for more formats.**
  ///
  /// You must add your icon to your Android project's `res/drawable` directory.
  /// Example: `android/app/src/main/res/drawable/notification_icon.png`
  ///
  /// And pass: `icon: notification_icon` without the file extension.
  ///
  /// If `null`, the default app icon will be used.
  /// Defaults to `null`.
  final String? icon;

  /// The color of the notification icon.
  ///
  /// **Only customizable for Android. On iOS, app default icon will be shown.**
  ///
  /// The icon is monochrome (light or dark) in the status bar but in expanded
  /// view it gets a background color that can be set with this parameter.
  ///
  /// If `null`, the icon will have a default color.
  /// Defaults to `null`.
  final Color? iconColor;

  /// Custom action buttons shown on the alarm notification.
  ///
  /// **Fork addition — Android only.** Each entry is rendered as a separate
  /// notification action. Tapping one stops the alarm and records its
  /// [NotificationActionButton.id], which the app drains via
  /// `Alarm.consumePendingNotificationActions()`.
  ///
  /// When non-empty this replaces [stopButton] on Android. Empty by default,
  /// which keeps the upstream single-[stopButton] behaviour.
  final List<NotificationActionButton> actionButtons;

  /// Converts the `NotificationSettings` instance to a JSON object.
  Map<String, dynamic> toJson() => _$NotificationSettingsToJson(this);

  /// Converts to wire datatype which is used for host platform communication.
  ///
  /// [actionButtons] is carried as a JSON-encoded string so the Pigeon wire
  /// type only needs a single extra (trailing, nullable) field.
  NotificationSettingsWire toWire() => NotificationSettingsWire(
        title: title,
        body: body,
        stopButton: stopButton,
        icon: icon,
        iconColorAlpha: iconColor?.a,
        iconColorRed: iconColor?.r,
        iconColorGreen: iconColor?.g,
        iconColorBlue: iconColor?.b,
        actionButtons: actionButtons.isEmpty
            ? null
            : jsonEncode(
                actionButtons.map((b) => b.toJson()).toList(growable: false),
              ),
      );

  /// Creates a copy of this notification settings but with the given fields
  /// replaced with the new values.
  NotificationSettings copyWith({
    String? title,
    String? body,
    String? stopButton,
    String? icon,
    Color? iconColor,
    List<NotificationActionButton>? actionButtons,
  }) {
    assert(title != null, 'NotificationSettings.title cannot be null');
    assert(body != null, 'NotificationSettings.body cannot be null');

    return NotificationSettings(
      title: title ?? this.title,
      body: body ?? this.body,
      stopButton: stopButton ?? this.stopButton,
      icon: icon ?? this.icon,
      iconColor: iconColor ?? this.iconColor,
      actionButtons: actionButtons ?? this.actionButtons,
    );
  }

  @override
  List<Object?> get props => [
        title,
        body,
        stopButton,
        icon,
        iconColor,
        actionButtons,
      ];
}
